### PR TITLE
Support promoted types in Quasi Quoter

### PR DESCRIPTION
**Pull request by danbroooks**
Before submitting your PR, check that you've:

- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [ ] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock
- [ ] Ran `stylish-haskell` on any changed files.
- [x] Adhered to the code style (see the `.editorconfig` file for details)

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [ ] Bumped the version number if there isn't an `(unreleased)` on the Changelog
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

--------

Closes https://github.com/yesodweb/persistent/issues/1166

I've opened this as a draft PR with a view towards getting some feedback on it still. I think everything here should be fine for addressing that ticket, but there might be something I have overlooked here. I will annotate what I've got so far.

I have also opened this PR against `master` as it is not a super invasive change, and I would be happy to deal with porting it forward into the `persistent-2.13` branch after it is merged in. Alternatively if it is preferred that this PR goes against that branch, and is backported to `master`, I'd be happy to do that as well.

It may also make sense for this to just go into `2.13` like my other recent change :+1:

### CODE REVIEW

File: persistent/test/Database/Persist/TH/KindEntitiesSpec.hs
Comment (danbroooks) [inline persistent/test/Database/Persist/TH/KindEntitiesSpec.hs:39]: There might be some more stuff to do here for this test maybe ... for now I have something that just asserts that it typechecks :+1:

File: persistent/Database/Persist/TH.hs
Comment (danbroooks) [inline persistent/Database/Persist/TH.hs:353]: Not sure about this. I basically just mirrored what is done for `FTTypeCon` when there is no module (ie the case above).

File: persistent/Database/Persist/Quasi.hs
Comment (parsonsmatt) [inline persistent/Database/Persist/Quasi.hs:511]: Looks great 👍🏻

File: persistent/Database/Persist/TH.hs
Comment (parsonsmatt) [reply persistent/Database/Persist/TH.hs:353]: Hmm - I'd say let's go with a `Left Nothing` on all cases here. If we hit this, it means we're trying to determine whether we need an embedded entity reference, which we shouldn't - the only place this could *possibly* be is if we've got an `FTApp` or similar, and we want the SQL type construction to proceed unchanged.

```suggestion
    Left Nothing
```

Like, we'll have `foo   (Proxy 'User)`. That'll get turned into `FTApp (FTypeCon Nothing "Proxy") (FTTypePromoted "User")`. This code recurses and then finds that the corresponding `EmbedEntityDef` is for a `User`, even though that's totally wrong.

File: persistent/Database/Persist/TH.hs
Comment (parsonsmatt) [inline persistent/Database/Persist/TH.hs:1777]: 👨🏻‍🍳

File: persistent/test/Database/Persist/TH/KindEntitiesSpec.hs
Comment (parsonsmatt) [inline persistent/test/Database/Persist/TH/KindEntitiesSpec.hs:16]: beautiful tests <3

File: persistent/test/Database/Persist/TH/KindEntitiesSpec.hs
Comment (parsonsmatt) [reply persistent/test/Database/Persist/TH/KindEntitiesSpec.hs:39]: Yeah - let's check the type of the field accessors and the `EntityField`s.

Still should jsut be a compilation check, but something like:

```haskell
let mkTransfer :: CustomerId -> MoneyAmount 'CustomerOwned 'Debt -> CustomerTransfer
    mkTransfer = CustomerTransfer
    getAmount :: CustomerTransfer -> MoneyAmount 'CustomerOwned 'Debt
    getAmount = customerTransferMoneyAmount
```

File: persistent/test/Database/Persist/TH/KindEntitiesSpecImports.hs
Comment (parsonsmatt) [inline persistent/test/Database/Persist/TH/KindEntitiesSpecImports.hs:18]: I don't think this `PeristRational` lines up with the `SqlInt64` type you specified

**Review by parsonsmatt — CHANGES_REQUESTED** [review]
There are a few changes I'd like on this before accepting it, but it is fantastic work - thank you so much!

I'm afraid the changes I put into #1256  are going to be a massive conflict here. So let's get this merged in to `master`, we can release it as a patch bump since it's just a change to the QQ and purely additive. Then I'll work on getting it merged into `persistent-2.13`.

**Review by parsonsmatt — APPROVED** [review]
🎉

**Comment by parsonsmatt** [comment]
All good to merge?

**Comment by danbroooks** [comment]
Yeah all good :+1: I'm happy to do the merge/conflict resolution into `persistent-2.13` if you want (I can do it straight after this PR is merged if it is going in now), I'm also happy to leave it to you :smile:

**Comment by parsonsmatt** [comment]
That would be really helpful, thank you 🙌🏻

**Pull request merged**
